pkg/deploy/controller/deployment/controller.go

```diff
@@ -141,14 +136,34 @@ func (c *DeploymentController) Handle(deployment *kapi.ReplicationController) er
 			glog.V(4).Infof("Cancelling %d deployer pods for deployment %s", len(deployerPods), deployutil.LabelForDeployment(deployment))
 			zeroDelay := int64(1)
 			for _, deployerPod := range deployerPods {
-				// Set the ActiveDeadlineSeconds on the pod so it's terminated very soon.
-				if deployerPod.Spec.ActiveDeadlineSeconds == nil || *deployerPod.Spec.ActiveDeadlineSeconds != zeroDelay {
-					deployerPod.Spec.ActiveDeadlineSeconds = &zeroDelay
-					if _, err := c.podClient.updatePod(deployerPod.Namespace, &deployerPod); err != nil {
-						c.recorder.Eventf(deployment, "failedCancellation", "Error cancelling deployer pod %s for deployment %s: %v", deployerPod.Name, deployutil.LabelForDeployment(deployment), err)
-						return fmt.Errorf("couldn't cancel deployer pod %s for deployment %s: %v", deployerPod.Name, deployutil.LabelForDeployment(deployment), err)
+				if deployerPod.Status.Phase == kapi.PodPending {
+					// If the pod hasn't yet been scheduled, just delete it.
+					err := c.podClient.deletePod(deployerPod.Namespace, deployerPod.Name)
+					if err != nil {
+						c.recorder.Eventf(deployment, "failedCancellation", "Error deleting deployer pod %s while cancelling deployment %s: %v", deployerPod.Name, deployutil.LabelForDeployment(deployment), err)
+						return fmt.Errorf("couldn't delete deployer pod %s while cancelling deployment %s: %v", deployerPod.Name, deployutil.LabelForDeployment(deployment), err)
+					}
+					// If we deleted the unscheduled primary deployer pod, just
+					// transition to failed immediately, since the DeployerPodController
+					// has no cleanup to do as the deployment never executed.
+					if deployerPod.Name == deployutil.DeployerPodNameForDeployment(deployment.Name) {
```

Comment:
I think cancel should probably be delete.  The historical reason was
probably logs, but that will have other solutions.  I don't think the
deadline gives us anything.  When graceful delete lands maybe we want this
to be a graceful deletion.

@derekwaynecarr do you remember why deadline was used instead of just
deleting?

On Thu, Aug 27, 2015 at 3:48 PM, Dan Mace <notifications@github.com> wrote:

> In pkg/deploy/controller/deployment/controller.go
> <https://github.com/openshift/origin/pull/4364#discussion_r38131918>:
>
> > -				if deployerPod.Spec.ActiveDeadlineSeconds == nil || *deployerPod.Spec.ActiveDeadlineSeconds != zeroDelay {
> > -					deployerPod.Spec.ActiveDeadlineSeconds = &zeroDelay
> > -					if _, err := c.podClient.updatePod(deployerPod.Namespace, &deployerPod); err != nil {
> > -						c.recorder.Eventf(deployment, "failedCancellation", "Error cancelling deployer pod %s for deployment %s: %v", deployerPod.Name, deployutil.LabelForDeployment(deployment), err)
> > -						return fmt.Errorf("couldn't cancel deployer pod %s for deployment %s: %v", deployerPod.Name, deployutil.LabelForDeployment(deployment), err)
> > +				if deployerPod.Status.Phase == kapi.PodPending {
> > +					// If the pod hasn't yet been scheduled, just delete it.
> > +					err := c.podClient.deletePod(deployerPod.Namespace, deployerPod.Name)
> > +					if err != nil {
> > +						c.recorder.Eventf(deployment, "failedCancellation", "Error deleting deployer pod %s while cancelling deployment %s: %v", deployerPod.Name, deployutil.LabelForDeployment(deployment), err)
> > +						return fmt.Errorf("couldn't delete deployer pod %s while cancelling deployment %s: %v", deployerPod.Name, deployutil.LabelForDeployment(deployment), err)
> > +					}
> > +					// If we deleted the unscheduled primary deployer pod, just
> > +					// transition to failed immediately, since the DeployerPodController
> > +					// has no cleanup to do as the deployment never executed.
> > +					if deployerPod.Name == deployutil.DeployerPodNameForDeployment(deployment.Name) {
>
> I can't remember now why deadline was chosen over delete. As long as it's
> safe to delete the pod no matter what state it's in (unscheduled, running,
> etc), I don't see why not just always try and delete.
>
> Also, we don't seem to set the state at the end of the loop to failed or
> something?
>
> I do that in this patch for the delete case: for the deadline case, the
> deployerpodcontroller will pick up the updated pod and handle the
> deployment state (when the deadline expires, the pod isn't deleted-
> containers are terminated and the pod is failed)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4364/files#r38131918>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift